File: 0053_maximum_subarray/maximum_subarray.kt
class Solution {
    fun maxSubArray(nums: IntArray): Int {
        var curSum = nums[0]
        return nums
            .asSequence()
            .drop(1)
            .fold(nums[0]) { acc, e ->
                curSum = maxOf(e, curSum + e)
                maxOf(acc, curSum)
            }
    }
}

File: 0053_maximum_subarray/maximum_subarray.ex
defmodule Solution do
  @spec max_sub_array(nums :: [integer]) :: integer
  def max_sub_array(nums) do
    f = nums |> Enum.at 0
    nums
        |> Enum.drop(1)
        |> Enum.reduce({f, f}, fn e, {acc, curSum} ->
                curSum = max curSum + e, e
                { max(acc, curSum), curSum }
            end)
        |> elem 0
  end
end
